### DISCUSSION TITLE
Builds failing because of downstream Matrix package dependency

### DISCUSSION THREAD

**Discussion by gavinsimpson** [r-universe-org/help#354] [Q&A] [answered]
Hi

Builds of my {gratia} package on Windows and MacOS are failing because of the incompatibility the Matrix package developers baked into their package late last year. gratia doesn't depend on Matrix, but it does have a dependency on gamm4, which depends on lme4, which depends on Matrix. And this is where the problems start. A representative failed run is shown here:

https://github.com/r-universe/gavinsimpson/actions/runs/7477709032/job/20351263645

I fixed this in my own GH Actions workflow by [inserting](https://github.com/gavinsimpson/gratia/commit/46d763fef13e2773666b3f70d2f924c626dbe4d3) (temporarily I hope) a step that installed the Matrix package into the runner of the workflow. This stopped it from using the version of Matrix that ships with R which is not compatible with lme4. For some reason the tool that finds dependencies in the r-libs action I am using couldn't identify the kind of dependency here and so updated versions of whichever of lme4, gamm4, or Matrix were not getting installed leading to the incompatibility. By forcing the action to install Matrix through the extra-package specification I could force installation during the build/check.

Is there a way to tell the r-universe builder to also do this so that builds on Windows and MacOS don't continue to fail?

The Matrix developers don't seem to be able to fix this issue (from what I read online around this),  as it doesn't seem to be an issue for users; you just reinstall lme4 and Matrix and things get fixed. But this isn't quite so easy on GH Actions and r-universe as I'm not in control of how dependencies are resolved.

And I do not want to force dependencies on lme4 and Matrix on my gratia package because it doesn't depend on these packages and nor should it.

Any help resolving this would be most appreciated.

**Comment by jeroen** [comment]
I think if you add `Suggests: Matrix` to your package, the latest version will automatically get installed before R CMD check.

**Reply by gavinsimpson** [reply]
Thanks; Maybe I should just do that, but it seems wrong to fix the mess they made by falsely claiming that my package does anything with Matrix.

**Reply by jeroen** [reply]
I guess the problem will resolve itself for the next version of R, when a newer Matrix is included?

**Reply by gavinsimpson** [reply]
Hopefully, but I don't know. I suppose add Matrix to Suggests is a small price to pay to get builds working again here.

**Reply by jeroen** [reply, marked as answer]
FWIW, the _builds_ on r-universe are working fine, binaries of your package are up to date. It is just a test that fails, but you are free to ignore that if you don't care.

**Reply by gavinsimpson** [reply]
Hmmm, something has changed recently then. For a while the binary builds were failing on Windows and MacOS (or they weren't in sync with the linux build). I didn't check as closely as I should when revisiting this and posting here. I saw the build failure and didn't check that the binaries are up to date.

I do care if tests fail; I don't care about this failure because of the Matrix issue as that is not something I can do anything about (and I know it's not an issue on user systems beyond the user needing to reinstall lme4 IIRC).

OK, thanks. I think the proximal concern (binaries not up to date with the linux build) is fixed now so I can ignore the Matrix failure thing.

Appreciate your input!

**Reply by jeroen** [reply]
Perhaps you can condition your failing unit test on the version of Matrix that is installed?
